res/controllers/common-controller-scripts.js

```diff
@@ -248,18 +256,12 @@ script.midiDebug = function(channel, control, value, status, group) {
 
 // Returns the deck number of a "ChannelN" or "SamplerN" group
 script.deckFromGroup = function(group) {
-    let deck = 0;
-    if (group.substring(2, 8) === "hannel") {
-        // Extract deck number from the group text
-        deck = group.substring(8, group.length - 1);
-    }
-    /*
-        else if (group.substring(2,8)=="ampler") {
-            // Extract sampler number from the group text
-            deck = group.substring(8,group.length-1);
-        }
-    */
-    return parseInt(deck);
+    try {
+        const deck = Number.parseInt(script.channelRegEx.exec(`${group}`)[1]);
+        return Number.isNaN(deck) ? 0 : deck;
+    } catch (_) {
+        return 0;
+    }
```

Comment:
Can you elaborate on why you prefer this over the previous implementation? I don't think [its worth the performance penalty](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJBUHRVQmxwajNNTXFPamZHU1ZEVGgiLCJjb2RlIjoiREFUQS5hcnIubWFwKERBVEEuZGVja0Zyb21Hcm91cFNsb3cpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6InpBUEZLTnhDcEZQSGVUckowVC0xVyIsImNvZGUiOiJEQVRBLmFyci5tYXAoREFUQS5kZWNrRnJvbUdyb3VwRmFzdCkiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6InJldHVybiB7XG4gIGNoYW5uZWxSZWdFeDogL15cXFtDaGFubmVsKFxcZCspXFxdJC8sXG4gIGFycjogQXJyYXkoMTAwMCkuZmlsbChcIltDaGFubmVsMV1cIiksXG4gIGRlY2tGcm9tR3JvdXBTbG93OiAoZ3JvdXApID0-IHtcbiAgdHJ5IHtcbiAgICAgICAgY29uc3QgZGVjayA9IE51bWJlci5wYXJzZUludChzY3JpcHQuY2hhbm5lbFJlZ0V4LmV4ZWMoYCR7Z3JvdXB9YClbMV0pO1xuICAgICAgICByZXR1cm4gTnVtYmVyLmlzTmFOKGRlY2spID8gMCA6IGRlY2s7XG4gICAgfSBjYXRjaCAoXykge1xuICAgICAgICByZXR1cm4gMDtcbiAgICB9XG59LFxuICBkZWNrRnJvbUdyb3VwRmFzdDogZ3JvdXAgPT4ge1xuICBsZXQgZGVjayA9IDA7XG4gICAgaWYgKGdyb3VwLnN1YnN0cmluZygyLCA4KSA9PT0gXCJoYW5uZWxcIikge1xuICAgICAgICAvLyBFeHRyYWN0IGRlY2sgbnVtYmVyIGZyb20gdGhlIGdyb3VwIHRleHRcbiAgICAgICAgZGVjayA9IGdyb3VwLnN1YnN0cmluZyg4LCBncm91cC5sZW5ndGggLSAxKTtcbiAgICB9XG4gIGRlY2sgPSBwYXJzZUludChkZWNrKTtcbiAgcmV0dXJuIE51bWJlci5pc05hTihkZWNrKSA_IDAgOiBkZWNrO1xufVxufVxuIn19). `deckFromGroup` is called far too often in the "simple" JS mappings so I want this to be fast.

Reply:
[The performance penalty is not that big](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJLUmEtSTFDMS1XdkNPcDFWTUpSYXoiLCJjb2RlIjoiREFUQS5hcnIubWFwKERBVEEuZGVja0Zyb21Hcm91cFNsb3cpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6IkFQdFVCbHBqM01NcU9qZkdTVkRUaCIsImNvZGUiOiJEQVRBLmFyci5tYXAoREFUQS5kZWNrRnJvbUdyb3VwU2xvdzIpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6InpBUEZLTnhDcEZQSGVUckowVC0xVyIsImNvZGUiOiJEQVRBLmFyci5tYXAoREFUQS5kZWNrRnJvbUdyb3VwRmFzdCkiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6InJldHVybiB7XG4gIGFycjogQXJyYXkoMTAwMCkuZmlsbChcIltDaGFubmVsMV1cIiksXG4gIGRlY2tGcm9tR3JvdXBTbG93OiAoZ3JvdXApID0-IHtcbiAgdHJ5IHtcbiAgICAgICAgY29uc3QgZGVjayA9IE51bWJlci5wYXJzZUludCgvXlxcW0NoYW5uZWwoXFxkKylcXF0kLy5leGVjKGAke2dyb3VwfWApWzFdKTtcbiAgICAgICAgcmV0dXJuIE51bWJlci5pc05hTihkZWNrKSA_IDAgOiBkZWNrO1xuICAgIH0gY2F0Y2ggKF8pIHtcbiAgICAgICAgcmV0dXJuIDA7XG4gICAgfVxufSxcbiAgZGVja0Zyb21Hcm91cFNsb3cyOiAoZ3JvdXApID0-IHtcbiAgICBjb25zdCBleHRyYWN0ZWQgPSAvXlxcW0NoYW5uZWwoXFxkKylcXF0kLy5leGVjKGAke2dyb3VwfWApO1xuICAgIGlmKGV4dHJhY3RlZCA9PT0gbnVsbCkgeyByZXR1cm4gMDsgfVxuICAgICAgICBjb25zdCBkZWNrID0gTnVtYmVyLnBhcnNlSW50KGV4dHJhY3RlZFsxXSk7XG4gICAgICAgIHJldHVybiBOdW1iZXIuaXNOYU4oZGVjaykgPyAwIDogZGVjaztcbn0sXG4gIGRlY2tGcm9tR3JvdXBGYXN0OiBncm91cCA9PiB7XG4gIGxldCBkZWNrID0gMDtcbiAgICBpZiAoZ3JvdXAuc3Vic3RyaW5nKDIsIDgpID09PSBcImhhbm5lbFwiKSB7XG4gICAgICAgIC8vIEV4dHJhY3QgZGVjayBudW1iZXIgZnJvbSB0aGUgZ3JvdXAgdGV4dFxuICAgICAgICBkZWNrID0gZ3JvdXAuc3Vic3RyaW5nKDgsIGdyb3VwLmxlbmd0aCAtIDEpO1xuICAgIH1cbiAgZGVjayA9IHBhcnNlSW50KGRlY2spO1xuICByZXR1cm4gTnVtYmVyLmlzTmFOKGRlY2spID8gMCA6IGRlY2s7XG59XG59XG4ifX0). `deckFromGroupSlow` had a typo (`script` is `undefined`).

Reply:
> deckFromGroupSlow had a typo (script is undefined) making the try to always endup in the catch clause.

good catch (pun intendended ;)

Anyways, [catching is still quite slow](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJLUmEtSTFDMS1XdkNPcDFWTUpSYXoiLCJjb2RlIjoiREFUQS5hcnIubWFwKERBVEEuZGVja0Zyb21Hcm91cFNsb3cpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6IkFQdFVCbHBqM01NcU9qZkdTVkRUaCIsImNvZGUiOiJEQVRBLmFyci5tYXAoREFUQS5kZWNrRnJvbUdyb3VwU2xvdzIpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6InpBUEZLTnhDcEZQSGVUckowVC0xVyIsImNvZGUiOiJEQVRBLmFyci5tYXAoREFUQS5kZWNrRnJvbUdyb3VwRmFzdCkiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6ImNvbnN0IHJlZyA9IC9eXFxbQ2hhbm5lbChcXGQrKVxcXSQvO1xucmV0dXJuIHtcbiAgYXJyOiBBcnJheSgxMDAwKS5maWxsKDApLm1hcCgoKSA9PiBNYXRoLnJhbmRvbSgpIDwgMC41ID8gXCJbQ2hhbm5lbDFdXCIgOiBcImdhcmJhZ2VcIiksXG4gIGRlY2tGcm9tR3JvdXBTbG93OiAoZ3JvdXApID0-IHtcbiAgdHJ5IHtcbiAgICAgICAgY29uc3QgZGVjayA9IE51bWJlci5wYXJzZUludChyZWcuZXhlYyhgJHtncm91cH1gKVsxXSk7XG4gICAgICAgIHJldHVybiBOdW1iZXIuaXNOYU4oZGVjaykgPyAwIDogZGVjaztcbiAgICB9IGNhdGNoIChfKSB7XG4gICAgICAgIHJldHVybiAwO1xuICAgIH1cbn0sXG4gIGRlY2tGcm9tR3JvdXBTbG93MjogKGdyb3VwKSA9PiB7XG4gICAgY29uc3QgZXh0cmFjdGVkID0gcmVnLmV4ZWMoYCR7Z3JvdXB9YCk7XG4gICAgaWYoZXh0cmFjdGVkID09PSBudWxsKSB7IHJldHVybiAwOyB9XG4gICAgICAgIGNvbnN0IGRlY2sgPSBOdW1iZXIucGFyc2VJbnQoZXh0cmFjdGVkWzFdKTtcbiAgICAgICAgcmV0dXJuIE51bWJlci5pc05hTihkZWNrKSA_IDAgOiBkZWNrO1xufSxcbiAgZGVja0Zyb21Hcm91cEZhc3Q6IGdyb3VwID0-IHtcbiAgbGV0IGRlY2sgPSAwO1xuICAgIGlmIChncm91cC5zdWJzdHJpbmcoMiwgOCkgPT09IFwiaGFubmVsXCIpIHtcbiAgICAgICAgLy8gRXh0cmFjdCBkZWNrIG51bWJlciBmcm9tIHRoZSBncm91cCB0ZXh0XG4gICAgICAgIGRlY2sgPSBncm91cC5zdWJzdHJpbmcoOCwgZ3JvdXAubGVuZ3RoIC0gMSk7XG4gICAgfVxuICBkZWNrID0gcGFyc2VJbnQoZGVjayk7XG4gIHJldHVybiBOdW1iZXIuaXNOYU4oZGVjaykgPyAwIDogZGVjaztcbn1cbn1cbiJ9fQ). So can we avoid that?

> Anyway, I find this version more resistent to passing stupid that may cause the controller to crash during a live (like undefined or -1, yes I did that developping my DN-S3700 controller).

Can you elaborate? Couldn't we gain the same resilience without having to reach for the regex?

Reply:
> So can we avoid that?

I'm ok with this.

> Can you elaborate? Couldn't we gain the same resilience without having to reach for the regex?

I don't remember exactly. I was probably working on something else and I decided to change that code while I was at it. I often do this kind of things. My development style is not structured at all.

Is there a guaranty that the group form will always be *[Channel1]* and not, for instance "[Channel 1]" or "[Channel 10]"? Generally I prefer regexes in this kind of use case since they tend to be more flexible and I'm not sure the overhead will have any significant impact.

But I'm ok, leaving this code as is. Although I'd still cast `group` to a string using `group = `${group}` to avoid stupid things like `deckFromGroup(undefined)` or `deckFromGroup(null)`.

Reply:
> My development style is mostly chaotic

No worries ;). Same goes for me. 

> Is there a guaranty that the group form will always be [Channel1] and not, for instance [Channel 1] or [Channel 12]?

Yes. That is pretty much set in stone for as long as CO's will exist.

> not sure the overhead will have any significant impact.

I'm not fully certain either, but its an easy and trivial optimization opportunity, judging by how often its called. 

> Although I'd still cast group to a string using group = `${group}` to avoid stupid things like deckFromGroup(undefined) or deckFromGroup(null).

Not sure I fully agree with that either. IMO nonsense code like that is supposed to crash, not just silently fail because `parseInt("undefined"[8])` is `NaN`. Much better to get a stacktrace right then and there. Wdyt?

Reply:
Alright, I'm fine with this. I reversed the change.